docs/internal/versioning.md

```diff
@@ -0,0 +1,244 @@
+ - Adding any top-level public field, method, or type.
+   This will break any file that imports the module into its top-level scope and defines something with the same name (or imports something else into its top-level scope that defines something with the same name).
+
+In most packages, it will be important to make this kind of breaking change periodically.
+The simplest way for a package author to do this is simply to bump the library version and require everyone to fix any breakages in their code.
+This is not desirable because the only way that this allows *graceful deprecation*, which provides a window in which old and new versions of the APIs exist, is to ship two versions of the library.
+Shipping two versions of the library causes several problems, most notably:
+
+ - Security fixes must be back-ported to the old code.
+   This is somewhat tractable when you have two versions but if this is needed for every API change then it can rapidly become dozens of versions.
+ - If a project uses package A and also package B that uses package A, then it is possible to end up with a dependency on two incompatible versions of package A.
+
+API Versioning
+--------------
+
+In Verona, we differentiate between the package (implementation) version and the API version.
+A package version is a simple monotonic counter of package releases.
+For each release, the package supports a set of API versions.
+API versions use semantic versioning, where the major version is incremented to indicate the removal of types or methods and the minor version to indicate the addition of new classes or methods.
+Both indicate potentially breaking changes for Verona consumers but it should be possible for a package to support all minor versions of an API with minimal effort by the package author.
+There is also one special API version: `unstable`.
+This represents any work-in-progress changes since the last release and exists so that developers can import the current snapshot of a library to test whether in-development changes are likely to cause problems.
+
+By making the API version distinct from the implementation version, we allow:
+
+ - A program to consume two different API versions of the same package (while consuming a single implementation)
+ - A package to deprecate and remove APIs gradually over time without needing two versions of the code.
+ - A package to add new APIs without breaking any consumers.
+
+When a package is imported, the string describing the package should include the API version, for example:
+
+```verona
+import "some/package@2.0" as p;
+```
+
+The build tooling should pick the most recent version of the package that supports all of the versions that are used in the project and its dependencies and warn if that is not the most recent available version.
+**Open question: Should pinning to specific implementations be part of this or should it be separate build metadata such as vcpkg.json?**
```

Comment:
I think there are two cases:
1. The code doesn't require a specific implementation but wants to avoid upgrade breakages. This is best done at a configuration file level.
2. The code can only work with a specific version. This is best done at the import level.

If the code imports multiple versions that are incompatible (either through config or `import`), the runtime should import both and separate them by symbol.

This could create the situation where the mangling is different after an upgrade, which would need recompilation of some parts that weren't necessarily part of the original upgrade (ie. other imports that moved to an intermediary API of the same library).

I think the migration path that ties API with code and data structure migration should take care of that, but the update can end up much larger than expected.

Reply:
Separating them by symbol in the presence of structural typing is not possible, at least in the general case.

Reply:
Right, as discussed, I was conflating library version with API version.

There's still the case where I only need an intersection of API versions 2 and 5, so don't care which API version I'm using on that range, but this is only problematic if we have to rely on third-party package distribution, where some OSs have the old packages and others the new ones. 

If we use our own package distribution (which comes with its own problems), we can always take the latest version.